### PR TITLE
Login and SignUp

### PR DESCRIPTION
**Pull request by omar-sarfraz**
- Added Postgres database connection
- Stored encrypted user data and hashed password in database upon signup
- Compare user records upon login

### CODE REVIEW

**Comment by omar-sarfraz** [comment]
- Added json web token and passport js

File: Backend/models/User.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import { DataTypes } from "sequelize";
+
+export const User = sequelize.define("User", {
+    id: {
+        type: DataTypes.INTEGER,
+        autoIncrement: true,
+        primaryKey: true,
+    },
+    firstName: {
+        type: DataTypes.STRING(50),
```
Comment (larebsyed) [inline Backend/models/User.ts:11]: there shouldn't be a limit on the name

File: Backend/models/User.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        allowNull: false,
+    },
+    email: {
+        type: DataTypes.STRING(150),
+        allowNull: false,
+        unique: true,
+    },
+    password: {
+        type: DataTypes.STRING(150),
```
Comment (larebsyed) [inline Backend/models/User.ts:24]: this should be password type field, so it is automatically saved as hash.

File: Backend/routes/auth/auth.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+const router = Router();
+
+router.post("/signup", signup);
+router.post("/login", login);
+
+// Demo protected route to test passport js functionality
+router.get(
+    "/secret",
+    passport.authenticate("jwt", { session: false }),
```
Comment (larebsyed) [inline Backend/routes/auth/auth.ts:22]: what does `session:false` do?

File: Backend/controllers/auth/signup.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        const encryptedFirstName = encrypt(key, userData.firstName);
+        const encryptedLastName = encrypt(key, userData.lastName);
+
+        const hash = await bcrypt.genSalt(key);
+        const encryptedPassword = await bcrypt.hash(userData.password, hash);
+
+        const user = await User.create({
+            firstName: encryptedFirstName,
```
Comment (larebsyed) [inline Backend/controllers/auth/signup.ts:40]: lets not encrypt first name and last name.

**Review by larebsyed — CHANGES_REQUESTED** [review]
Overall looks good, just suggested some improvements.

**Review by larebsyed — CHANGES_REQUESTED** [review]
Overall looks good, just suggested some improvements.

File: Backend/models/User.ts (hunk 1 above)
Comment (omar-sarfraz) [reply Backend/models/User.ts:11]: Should I use a TEXT type here?

File: Backend/routes/auth/auth.ts (hunk 3 above)
Comment (omar-sarfraz) [reply Backend/routes/auth/auth.ts:22]: Password js by default tries to create a session for user after authenticating. It is recommended to set it to false if we don't create persistent server side sessions for users.

File: Backend/controllers/auth/signup.ts (hunk 4 above)
Comment (omar-sarfraz) [reply Backend/controllers/auth/signup.ts:40]: Alright

File: Backend/models/User.ts (hunk 1 above)
Comment (omar-sarfraz) [reply Backend/models/User.ts]: I've used STRING type which by default creates a field in Postgres with VARCHAR(255).

**Pull request merged**